js/style/style_layer/fill_style_layer.js

```diff
@@ -7,6 +7,48 @@ function FillStyleLayer() {
     StyleLayer.apply(this, arguments);
 }
 
-module.exports = FillStyleLayer;
+FillStyleLayer.prototype = util.inherit(StyleLayer, {
+
+    getPaintValue: function(name, globalProperties, featureProperties) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.getPaintValue.call(this, 'fill-color', globalProperties, featureProperties);
+        } else {
+            return StyleLayer.prototype.getPaintValue.call(this, name, globalProperties, featureProperties);
+        }
+    },
+
+    getPaintValueStopZoomLevels: function(name) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.getPaintValueStopZoomLevels.call(this, 'fill-color');
+        } else {
+            return StyleLayer.prototype.getPaintValueStopZoomLevels.call(this, arguments);
+        }
+    },
+
+    getPaintInterpolationT: function(name, zoom) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.getPaintInterpolationT.call(this, 'fill-color', zoom);
+        } else {
+            return StyleLayer.prototype.getPaintInterpolationT.call(this, name, zoom);
+        }
+    },
 
-FillStyleLayer.prototype = util.inherit(StyleLayer, {});
+    isPaintValueFeatureConstant: function(name) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.isPaintValueFeatureConstant.call(this, 'fill-color');
+        } else {
+            return StyleLayer.prototype.isPaintValueFeatureConstant.call(this, name);
+        }
+    },
+
+    isPaintValueZoomConstant: function(name) {
+        if (name === 'fill-outline-color' && this.getPaintProperty('fill-outline-color') === undefined) {
+            return StyleLayer.prototype.isPaintValueZoomConstant.call(this, 'fill-color');
+        } else {
+            return StyleLayer.prototype.isPaintValueZoomConstant.call(this, name);
+        }
+    }
+
+});
```

Comment:
The proposed refactor sounds great! Yeah, we can follow-up on that later. I'd make sure to run benchmarks before merging though because things like this make things slower if they're in hot path.